libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs

```diff
@@ -355,63 +360,77 @@ instance NFData CostModelError
 -- of Plutus, which we cannot yet even validate. These are stored in
 -- 'costModelsUnknown`.
 data CostModels = CostModels
-  { costModelsValid :: !(Map Language CostModel)
-  , costModelsErrors :: !(Map Language CostModelError)
-  , costModelsUnknown :: !(Map Word8 [Integer])
+  { _costModelsValid :: !(Map Language CostModel)
+  , _costModelsErrors :: !(Map Language CostModelError)
+  , _costModelsUnknown :: !(Map Word8 [Integer])
   }
   deriving stock (Eq, Ord, Show, Generic)
 
+costModelsValid :: CostModels -> Map Language CostModel
+costModelsValid = _costModelsValid
+
+costModelsErrors :: CostModels -> Map Language CostModelError
+costModelsErrors = _costModelsErrors
+
+costModelsUnknown :: CostModels -> Map Word8 [Integer]
+costModelsUnknown = _costModelsUnknown
+
 emptyCostModels :: CostModels
 emptyCostModels = CostModels mempty mempty mempty
 
--- | Updates the first @CostModels@ with the second one so that only the
--- cost models that are present in the second one get updated while all the
--- others stay unchanged
+-- | Updates the first @CostModels@ with the second one so that only the cost models that
+-- are present in the second one get updated while all the others stay unchanged. Language
+-- specific errors are removed, whenever a valid CostModel for the language is updated.
 updateCostModels :: CostModels -> CostModels -> CostModels
-updateCostModels
-  (CostModels oldValid oldErrors oldUnk)
-  (CostModels newValid newErrors newUnk) =
-    CostModels
-      (Map.union newValid oldValid)
-      (Map.union newErrors oldErrors)
-      (Map.union newUnk oldUnk)
-
--- | This function attempts to add a new cost model to a given 'CostModels'.
--- If it is a valid cost model for a known version of Plutus, it is added to
--- 'validCostModels'. If it is an invalid cost model for a known version of Plutus,
--- the error is sorted in 'costModelErrors' and the cost model is stored in
--- 'invalidCostModels'. Lastly, if the Plutus version is unknown,
--- the cost model is also stored in 'invalidCostModels'.
-addRawCostModel :: Word8 -> [Integer] -> CostModels -> CostModels
-addRawCostModel langW8 cmIds (CostModels validCMs errs invalidCMs) =
-  case mkLanguageEnum (fromIntegral langW8) of
-    Just lang ->
-      case mkCostModel lang cmIds of
-        Right cm -> CostModels (Map.insert lang cm validCMs) errs invalidCMs
-        Left e -> CostModels validCMs (addError lang e errs) updatedInvalidCMs
-    Nothing -> CostModels validCMs errs updatedInvalidCMs
+updateCostModels (CostModels oldValid oldErrors oldUnk) (CostModels modValid modErrors modUnk) =
+  CostModels
+    newValid
+    (newErrors Map.\\ newValid)
+    (Map.union modUnk oldUnk Map.\\ Map.mapKeys languageToWord8 newValid)
```

Comment:
Nice, so this keeps the property that a Language is either in the valid or errors map that we have when using `mkCostModelsLenient`.   The resulting unknown costmodels  could still have keys that are in `errors`  (which is not possible in `mkCostModelsLenient`), but  don't think it matters.

Reply:
> The resulting unknown costmodels could still have keys that are in errors (which is not possible in mkCostModelsLenient), but don't think it matters.

That is actually necessary for rountripping. If we were to remove costmodels that caused errors from unknown then the errors field would be lost, since we do not serialize errors.